### ISSUE TITLE
cascade_deletes should also cascade restore

### ISSUE THREAD

**Issue by scott1702** [silverstripe/silverstripe-versioned#231]
## Affected Version

4.3

## Description

When adding dependant objects to cascade_deletes to a dataobject. If the dataobject is deleted and then restored, the objects which where cascade deleted do not get restored

## Steps to Reproduce

1. Create a page
1. Add a has_one or has_many relationship to a dataobject e.g. 'MyItem'
1. Add a `cascade_deletes` entry for 'MyItem'
1. Archive the page
1. Restore the page

Current behaviour: 
'MyItem' remains deleted

Expected behaviour:
'MyItem' gets restored along with the page

**Comment by robbieaverill** [comment]
I think the `cascade_deletes` API should also inherently support cascading restore actions, rather than adding a new API for it.

Keen to get @tractorcow's thoughts on this one.

**Comment by tractorcow** [comment]
Actually, I had in mind that cascade_deletes was the spritual opposite of `owns`.

cascade_deletes: These items should be removed with this item (negative requirement)
owns: These items must be added with this item (positive requirement)

The place where these differ are with shared items (i.e. items that don't belong to an exclusive parent), which are in owns, but not cascade_deletes.

The items that are in cascade_deletes but not owns is another part of the venn diagram... items that aren't strictly required by the parent (e.g. metadata). The difference is that these items rely on the parent, rather than the other way around.

In this sense, I would say a soft yes to the question "should items in cascade_deletes, but not owns, be restored?", but if not, I would argue you could just use `owns` as a reasonable alternative.

One could argue that you should only use cascade_deletes for determining removal, and owns for cascade adding, and that reversing this could have risks. Someone look into that a bit more. :)

So all in all, I'm like 70% in agreement, 30% some reservation. It's really just an opinion call really.

However, that would beg the question; Should there be a difference between "restore an object" and "restore an object with it's dependencies" at the code level? With publishing we have publishSingle() and publishRecursive(). If we did, the question becomes less of an issue, since you can just use restoreSingle() and manually restore the sub-components you need. :P

**Comment by sminnee** [comment]
A good example of why this is important is user forms. If you archive a user forms page and then restore it, currently you are forced to manually restore each form field, dropdown value of a dropdown, etc etc. It's extremely tedious.

In situations where the restore function is being used because the original delete was an accident and someone is trying to rapidly restore critical website functionality, doubly so.

Bumping priority to high because of this. ;-)

**Comment by sminnee** [comment]
On @tractorcow's point I see there are being two kinds of related content:

 * Non-exclusive composition: `owns`. E.g. "I show this file on my page"
 * Exclusive composition: `owns` + `cascade_deletes`. This is a more intuitive notion of ownerships, where the related data is part of this parent object and doesn't exist without it. E.g. windows in my house, or formfields on my userform.

I haven't really given much thought to what cascade_deletes but not owns would be. Probably owned, but non-content data, like the submissions of a userform or something.

----

One edge-case that needs to be considered with automatic restoration:

 * I have a userform with fields A, B, C
 * I delete field C
 * I delete the userform
 * I restore the userform

The required outcome in this case is that the userform, and fields A & B are restored, but not field C. Possibly this can be done by restoring the *latest version of the userform* and so hopefully it can build on recent work in this space.

**Comment by robbieaverill** [comment]
I think that last example wouldn't behave very nicely right now (see https://github.com/silverstripe/silverstripe-versioned/issues/195). Might be improved naturally by Aaron's work on versioned snapshots for that issue

**Comment by sminnee** [comment]
Yeah - versioned snapshots is critical. I get that it's a module because the content model changes are a bit experimental, but we shouldn't think of it as an optional extra.

**Comment by thisisannie** [comment]
Still an issue in CMS6, but it also affects unpublishing.

1. If parent and child objects are versioned (with cascade_deletes), unpublishing the parent will result in the child being unpublished. This is fine.
2. If child is _not_ versioned, cascade_deletes will result in the child record being deleted if the parent is unpublished. This results in unexpected data loss.

So this problem is that cascade_deletes for versioned objects is currently coupled to the parent_live table, not the parent table itself.

In older sites, database bloat becomes a real problem so it would help if we could use cascade_deletes where appropriate.

**Comment by michalkleiner** [comment]
Yeah, we'd probably want to add some logic to check whether the nested relations are versioned or not, and in the case of unpublishing the parent of a non-versioned child, probably leave the child be. Or, in a new minor version add a config to allow changing the behaviour. Then deciding what should be the default for the next major. Any thoughts on this @GuySartorelli?

**Comment by GuySartorelli** [comment]
Honestly I think we probably need to look at the way relationships are managed at a pretty high level and rework a lot of stuff (I talk about this a bit in https://github.com/silverstripe/silverstripe-elemental/issues/804 as well). The published/unpublished scenario is (theoretically) pretty straight forward - but Sam brings up some good points about how archived/restore could aggressively restore more than you intended it to.

Along with that, because `has_many` data on the parent actually gets stored as `has_one` on the child, restoring child records for that relation would be difficult.

There's also not a lot of fidelity with the `cascade_deletes` (which also cascades unpublish and archive events) and `owns` (which cascades publish and... honestly who knows what else). Even the fact that there's uncertainty here about which of those should control cascading a restore points at needing some thought into possibly reworking things.

Plus there's [confusion about whether the end of the relationship or the join record should be removed](https://github.com/silverstripe/silverstripe-framework/issues/9612) for `many`-type relations

And then when we go beyond the configuration.... the way versioned is tacked onto the ORM sometimes feels a little clunky - for example calling `write()` on the draft and then live version for a save-and-publish event means developers have to be really careful about their onbefore/onafter write hooks. I get why it's done that way, we don't want to add versioning logic or coupling directly in framework, but I would still like to look at reworking it some day to be more straight forward.

I dunno when we'd get the chance to look into all that though...

## To your specific comment

> we'd probably want to add some logic to check whether the nested relations are versioned or not, and in the case of unpublishing the parent of a non-versioned child, probably leave the child be.

If we just skip unversioned records, and they have a valid URL route to get to them separately, then we have unpublished parents whose children are still accessible on the frontend, which is probably undesirable.

One solution would be to make it easy to remove the `save` button for versioned objects and recommend versioning everything. The only button then is "save and publish", and unpublishing the parent only _unpublishes_ the child, not deletes it.
Of course versioning everything has its own challenges, with increased database size, more complexity, [having to publish everything on first deploy](https://github.com/silverstripe/silverstripe-versioned/issues/215), etc. This is more of a clunky workaround than a solution IMO.

There's also some discussion about this in https://github.com/silverstripe/silverstripe-versioned/issues/219 where sam said the following which is in line with your thinking:
> The simplest solution is probably for cascade_deletes from a versioned object to a non-versioned object to only delete the child object when the draft version is deleted. This would mean that changes to the child object are instantly reflected on the live site. However, typically a page being archived is an action that affects live & draft simultaneously, so that shouldn't be an issue.
>
> That said, I would probably also recommend a warning in dev/build or somewhere that we don't recommend using cascade_deletes from a versioned to a non-versioned object. We're never going to create a solution that doesn't have some weird edge-cases, unless everything is versioned.

The question then is whether that should just _happen_ in a minor release or whether it should be opt-in until CMS 7. But I think deleting the record _entirely_ just because you unpublished the parent is probably an outright bug, so I think I'd be okay with the change happening in a minor and being called out clearly in the changelog.

If we wanted to add a warning like Sam talked about, it could be part of the relation validation service.

**Comment by GuySartorelli** [comment]
All of that said - it's probably best for the "don't delete child unversioned records when unpublishing a parent versioned record" work to be done in a separate issue (probably just link here from https://github.com/silverstripe/silverstripe-versioned/issues/219 and use that issue).

It's related to what this issue discusses, but isn't exactly the same and can be handled without having to figure out what to do about the archive/restore situation.

**Comment by sminnee** [comment]
As mentioned in silverstripe/silverstripe-framework#9612, one approach would be that `$owns` implies both `$cascade_deletes` and `$cascade_restores`, and normal users are recommended to set `$owns`, not the specific cascade properties.

as a v7 / major release change
